File: build-judge-system/environment/problems/reorder/checker.cpp
#include "testlib.h"
#include <vector>
#include <algorithm>
using namespace std;

int main(int argc, char* argv[]) {
    registerTestlibCmd(argc, argv);

    int n = inf.readInt();
    vector<int> a(n);
    for (int i = 0; i < n; i++) a[i] = inf.readInt();

    vector<int> result;
    for (int i = 0; i < n; i++) {
        result.push_back(ouf.readInt());
    }

    // Must be a permutation of the input array
    vector<int> sa = a, sr = result;
    sort(sa.begin(), sa.end());
    sort(sr.begin(), sr.end());
    if (sa != sr)
        quitf(_wa, "output is not a permutation of input");

    // Must be a derangement: no element at its original position
    for (int i = 0; i < n; i++) {
        if (result[i] == a[i])
            quitf(_wa, "position %d has same element as input", i + 1);
    }

    quitf(_ok, "valid derangement of %d elements", n);
}

File: build-judge-system/environment/problems/reorder/checker.py
#!/usr/bin/env python3
"""
Custom checker for the reorder problem.
Validates that the output is a derangement of the input array:
  1. Output must be a permutation of the input elements.
  2. No element may remain at its original position.

Usage: checker.py <input_file> <expected_output_file> <actual_output_file>
Exit code 0 = accepted, non-zero = wrong answer.
"""
import sys


def main():
    input_file = sys.argv[1]
    _expected_file = sys.argv[2]
    actual_file = sys.argv[3]

    with open(input_file) as f:
        lines = f.read().strip().split('\n')
        n = int(lines[0])
        original = list(map(int, lines[1].split()))

    with open(actual_file) as f:
        content = f.read().strip()
        if not content:
            sys.exit(1)
        try:
            result = list(map(int, content.split()))
        except ValueError:
            sys.exit(1)

    if len(result) != n:
        sys.exit(1)

    if sorted(result) != sorted(original):
        sys.exit(1)

    for i in range(n):
        if result[i] == original[i]:
            sys.exit(1)

    sys.exit(0)


if __name__ == '__main__':
    main()
